File: javascript_basics/src/tests/dynamic_controls.test.js
require("../testSetup.js");
const { By, until } = require("selenium-webdriver");
const { getDriver } = require("../globalSetup.js");
const { logger } = require("../utils/logger.js");
const { sleep } = require("../utils/sleep.js");

describe("Dynamic Controls", function () {
  it("should remove and add the checkbox and enable/disable input field", async function () {

    const driver = getDriver();
    logger.info("STARTING Test --- Dynamic Controls ---");

    const url = "https://the-internet.herokuapp.com/dynamic_controls";
    logger.info(`Navigating to ${url}`);
    await driver.get(url);
    // await sleep(1000);

    logger.info("Locating Remove button");
    const removeButton = await driver.findElement(By.xpath("//button[text()='Remove']"));
    // await sleep(1000);

    logger.info("Clicking Remove button");
    await removeButton.click();
    // await sleep(1000);

    logger.info("Waiting for checkbox to be removed");
    await driver.wait(until.elementLocated(By.id("message")), 5000);
    const removeMessage = await driver.findElement(By.id("message")).getText();
    // await sleep(1000);

    if (!removeMessage.includes("gone")) {
      throw new Error(`Expected 'gone' message, but got: ${removeMessage}`);
    }

    logger.info("Clicking Add button");
    const addButton = await driver.findElement(By.xpath("//button[text()='Add']"));
    await addButton.click();
    // await sleep(1000);

    logger.info("Waiting for checkbox to return");
    await driver.wait(until.elementLocated(By.id("message")), 5000);
    const addMessage = await driver.findElement(By.id("message")).getText();
    // await sleep(1000);

    if (!addMessage.includes("back")) {
      throw new Error(`Expected 'back' message, but got: ${addMessage}`);
    }

    logger.info("Locating Enable button");
    const enableButton = await driver.findElement(By.xpath("//button[text()='Enable']"));
    // await sleep(1000);

    logger.info("Clicking Enable button");
    await enableButton.click();
    // await sleep(1000);

    logger.info("Waiting for input to be enabled");
    await driver.wait(until.elementLocated(By.id("message")), 5000);
    const enableMessage = await driver.findElement(By.id("message"));
    await driver.wait(until.elementTextIs(enableMessage, "It's enabled!"), 5000);
    // await sleep(1000);

    const inputField = await driver.findElement(By.css("input[type='text']"));
    const enabled = await inputField.isEnabled();
    // await sleep(1000);

    if (!enabled) {
      throw new Error("Input field was not enabled.");
    }

    logger.info("Locating Disable button");
    const disableButton = await driver.findElement(By.xpath("//button[text()='Disable']"));
    // await sleep(1000);

    logger.info("Clicking Disable button");
    await disableButton.click();
    // await sleep(1000);

    logger.info("Waiting for input to be disabled");
    await driver.wait(until.elementLocated(By.id("message")), 5000);
    const disableMessage = await driver.findElement(By.id("message"));
    await driver.wait(until.elementTextIs(disableMessage, "It's disabled!"), 5000);
    // await sleep(1000);

    const disabled = !(await inputField.isEnabled());
    if (!disabled) {
      throw new Error("Input field was not disabled.");
    }

    logger.info("ENDING Test --- Dynamic Controls ---");
  });
});

File: csharp_basics/tests/dynamic_controls_test.cs
using System;
using NUnit.Framework;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using csharp_basics.utils;

namespace csharp_basics.tests
{
    [TestFixture]
    public class DynamicControlsTest : BaseTest
    {
        [Test]
        public void TestDynamicControls()
        {
            Logger.Write("STARTING Test --- Dynamic Controls ---");

            string url = "https://the-internet.herokuapp.com/dynamic_controls";
            driver.Navigate().GoToUrl(url);
            Logger.Write($"Navigated to {url}");

            var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));

            // Remove Checkbox
            Logger.Write("Clicking 'Remove' button to remove checkbox");
            IWebElement removeButton = driver.FindElement(By.CssSelector("button[onclick='swapCheckbox()']"));
            removeButton.Click();

            Logger.Write("Waiting for checkbox to be removed");
            wait.Until(d =>
            {
                try
                {
                    var el = d.FindElement(By.Id("checkbox"));
                    return !el.Displayed;
                }
                catch (NoSuchElementException)
                {
                    return true; // element gone
                }
            });

            string message = driver.FindElement(By.Id("message")).Text;
            Logger.Write($"Message after removal: {message}");
            Assert.That(message, Does.Contain("It's gone!"));

            // Add Checkbox
            Logger.Write("Clicking 'Add' button to add checkbox back");
            IWebElement addButton = driver.FindElement(By.CssSelector("button[onclick='swapCheckbox()']"));
            addButton.Click();

            Logger.Write("Waiting for checkbox to reappear");
            wait.Until(d =>
            {
                try
                {
                    var el = d.FindElement(By.Id("checkbox"));
                    return el.Displayed;
                }
                catch (NoSuchElementException)
                {
                    return false;
                }
            });

            message = driver.FindElement(By.Id("message")).Text;
            Logger.Write($"Message after adding: {message}");
            Assert.That(message, Does.Contain("It's back!"));

            // Enable input field
            Logger.Write("Clicking 'Enable' button to activate input field");
            IWebElement enableButton = driver.FindElement(By.CssSelector("button[onclick='swapInput()']"));
            enableButton.Click();

            Logger.Write("Waiting for input field to become enabled");
            By inputLocator = By.CssSelector("#input-example input");
            wait.Until(d =>
            {
                var el = d.FindElement(inputLocator);
                return el.Displayed && el.Enabled;
            });

            message = driver.FindElement(By.Id("message")).Text;
            Logger.Write($"Message after enabling: {message}");
            Assert.That(message, Does.Contain("It's enabled!"));

            // Disable input field
            Logger.Write("Clicking 'Disable' button to deactivate input field");
            IWebElement disableButton = driver.FindElement(By.CssSelector("button[onclick='swapInput()']"));
            disableButton.Click();

            Logger.Write("Waiting for input field to become disabled");
            wait.Until(d =>
            {
                var el = d.FindElement(inputLocator);
                return !el.Enabled;
            });

            message = driver.FindElement(By.Id("message")).Text;
            Logger.Write($"Message after disabling: {message}");
            Assert.That(message, Does.Contain("It's disabled!"));

            Logger.Write("ENDING Test --- Completed successfully: Dynamic Controls ---");
        }
    }
}
